src/software/thunderscope/binary_context_managers.py

```diff
@@ -137,45 +137,43 @@ def setup_proto_unix_io(self, proto_unix_io):
 
         # Setup LOG(VISUALIZE) handling from full system. We set from_log_visualize
         # to true to decode from base64.
-        for arg in [
-            (self.full_system_runtime_dir, Obstacles, True),
-            (self.full_system_runtime_dir, PathVisualization, True),
-            (self.full_system_runtime_dir, PassVisualization, True),
-            (self.full_system_runtime_dir, NamedValue, True),
-            (self.full_system_runtime_dir, PlayInfo, True),
+        for proto_class in [
+            Obstacles,
+            PathVisualization,
+            PassVisualization,
+            NamedValue,
+            PlayInfo,
         ]:
-            proto_unix_io.attach_unix_receiver(*arg)
+            proto_unix_io.attach_unix_receiver(
+                runtime_dir=self.full_system_runtime_dir,
+                proto_class=proto_class,
+                from_log_visualize=True,
+            )
 
         proto_unix_io.attach_unix_receiver(
-            self.full_system_runtime_dir + "/log", RobotLog
+            self.full_system_runtime_dir, "/log", RobotLog
         )
 
-        # Inputs to full_system
-        for arg in [
-            (self.full_system_runtime_dir + ROBOT_STATUS_PATH, RobotStatus),
-            (self.full_system_runtime_dir + SSL_WRAPPER_PATH, SSL_WrapperPacket),
-            (self.full_system_runtime_dir + SSL_REFEREE_PATH, Referee),
-            (self.full_system_runtime_dir + SENSOR_PROTO_PATH, SensorProto),
-            (
-                self.full_system_runtime_dir + TACTIC_OVERRIDE_PATH,
-                AssignedTacticPlayControlParams,
-            ),
-            (self.full_system_runtime_dir + PLAY_OVERRIDE_PATH, Play),
-            (
-                self.full_system_runtime_dir + DYNAMIC_PARAMETER_UPDATE_REQUEST_PATH,
-                ThunderbotsConfig,
-            ),
-        ]:
-            proto_unix_io.attach_unix_sender(*arg)
-
         # Outputs from full_system
         proto_unix_io.attach_unix_receiver(
-            self.full_system_runtime_dir + WORLD_PATH, World
+            self.full_system_runtime_dir, WORLD_PATH, World
         )
         proto_unix_io.attach_unix_receiver(
-            self.full_system_runtime_dir + PRIMITIVE_PATH, PrimitiveSet
+            self.full_system_runtime_dir, PRIMITIVE_PATH, PrimitiveSet
         )
 
+        # Inputs to full_system
+        for arg in [
+            (ROBOT_STATUS_PATH, RobotStatus),
+            (SSL_WRAPPER_PATH, SSL_WrapperPacket),
+            (SSL_REFEREE_PATH, Referee),
+            (SENSOR_PROTO_PATH, SensorProto),
+            (TACTIC_OVERRIDE_PATH, AssignedTacticPlayControlParams,),
+            (PLAY_OVERRIDE_PATH, Play),
+            (DYNAMIC_PARAMETER_UPDATE_REQUEST_PATH, ThunderbotsConfig,),
+        ]:
+            proto_unix_io.attach_unix_sender(self.full_system_runtime_dir, *arg)
```

Comment:
clean :sparkles:

I wish it I wrote it like this to begin with, nice work :100: